src/solver/bounded.rs

```diff
@@ -0,0 +1,1429 @@
+            state.get_mut(&self.set).unwrap().insert(self.element);
+        }
+    }
+}
+
+use im::{vector, Vector};
+type Trace = Vector<State>;
+
+/// A bounded model checker will either find a counterexample or else not tell us anything
+#[derive(Debug, PartialEq)]
+pub enum InterpreterResult {
+    /// The checker found a counterexample, here it is
+    Counterexample(Trace),
+    /// The checker could not find any counterexamples
+    Unknown,
+}
+
+/// Run a given Program out to some depth
+/// Note that max_depth refers to the number of timesteps,
+/// e.g. a max_depth of 0 means only evaluate the initial states
+#[allow(dead_code)]
+pub fn interpret(program: &Program, max_depth: usize) -> InterpreterResult {
+    use std::collections::{HashSet as Cache, VecDeque as Queue};
+    let mut queue: Queue<Trace> = program
+        .inits
+        .iter()
+        .map(|state| vector![state.clone()])
+        .collect();
+    // cache stores states that have ever been present in the queue
+    let mut cache: Cache<State> = program.inits.iter().cloned().collect();
+
+    while let Some(trace) = queue.pop_front() {
+        let state = trace.last().unwrap();
+        if !program
+            .safes
+            .iter()
+            .any(|guards| guards.iter().all(|guard| guard.run(state)))
+        {
+            return InterpreterResult::Counterexample(trace);
+        }
+        if trace.len() - 1 < max_depth {
+            for tr in &program.trs {
+                if tr.guards.iter().all(|guard| guard.run(state)) {
+                    let mut next = state.clone();
+                    tr.updates.iter().for_each(|update| update.run(&mut next));
+                    if !cache.contains(&next) {
+                        cache.insert(next.clone());
+                        let mut trace = trace.clone();
+                        trace.push_back(next);
+                        queue.push_back(trace);
+                    }
+                }
+            }
+        }
+    }
+
+    InterpreterResult::Unknown
+}
+
+#[allow(missing_docs)]
+#[derive(Error, Debug, PartialEq)]
+pub enum TranslationError {
+    #[error("sort checking error: {0}")]
+    SortError(SortError),
+    #[error("sort {0} not found in universe {1:#?}")]
+    UnknownSort(String, HashMap<String, usize>),
+    #[error("all assumes should precede all asserts, but found {0:?}")]
+    OutOfOrderStatement(ThmStmt),
+    #[error("found an assert that isn't a safety property")]
+    AssertWithoutAlways(Term),
+    #[error("unknown identifier {0}")]
+    UnknownId(String),
+    #[error("can't translate term ({0}) with context {1:#?}")]
+    UnsupportedTerm(Term, Map<String, usize>),
+    #[error("could not reduce term {0:#?} to an element")]
+    NoValue(Valued),
+    #[error("we don't support negating no-ops, but found {0:?}")]
+    NegatedNoOp(Valued),
+    #[error("expected and: {0:#?}")]
+    ExpectedAnd(Valued),
+    #[error("expected or: {0:#?}")]
+    ExpectedOr(Valued),
```

Comment:
When I've seen these, it's been because something has simplified out and we end up with a literal true or false. I don't think that they can renamed on the Rust side, but I'll add that to the error messages.